chips/sam4l/src/usart.rs

```diff
@@ -586,6 +586,28 @@ impl USART {
             self.disable_tx_empty_interrupt(usart);
             self.disable_tx(usart);
             self.usart_tx_state.set(USARTStateTX::Idle);
+
+            // Now that we know the TX transaction is finished we can get the
+            // buffer back from DMA and pass it back to the client. If we don't
+            // wait until we are completely finished, then the
+            // `transmit_complete` callback is in a "bad" part of the USART
+            // state machine, and clients cannot issue other USART calls from
+            // the callback.
+            let buffer = self.tx_dma.get().map_or(None, |tx_dma| {
+                let buf = tx_dma.abort_xfer();
+                tx_dma.disable();
+                buf
+            });
+
+            // alert client
+            self.client.get().map(|usartclient| {
+                buffer.map(|buf| match usartclient {
+                    UsartClient::Uart(client) => {
+                        client.transmit_complete(buf, hil::uart::Error::CommandComplete);
+                    }
+                    UsartClient::SpiMaster(_) => {}
+                });
+            });
```

Comment:
Simplify code by replacing `match usartclient` with `if let UsartClient::Uart(client)` since you only care about one pattern!